File: src/preview/preview_geometry.c
#include "preview_geometry.h"

#include <math.h>

static double shaula_clamp_double(double value, double minimum,
                                  double maximum) {
  if (value < minimum)
    return minimum;
  if (value > maximum)
    return maximum;
  return value;
}

static unsigned int shaula_color_channel(double value) {
  /* Avoid undefined float-to-integer conversion for malformed NaN input. */
  if (isnan(value))
    return 0;
  double clamped = shaula_clamp_double(value, 0.0, 1.0);
  return (unsigned int)(clamped * 255.0 + 0.5);
}

ShaulaColor shaula_color_default(void) {
  return (ShaulaColor){0.165, 0.290, 0.400, 1.0};
}

void shaula_color_to_hex(ShaulaColor color, char out[8]) {
  g_return_if_fail(out != NULL);

  g_snprintf(out, 8, "#%02X%02X%02X", shaula_color_channel(color.r),
             shaula_color_channel(color.g), shaula_color_channel(color.b));
}

ShaulaRect shaula_rect_from_points(ShaulaPoint a, ShaulaPoint b) {
  return shaula_rect_normalized((ShaulaRect){a.x, a.y, b.x - a.x, b.y - a.y});
}

ShaulaRect shaula_rect_normalized(ShaulaRect rect) {
  if (rect.width < 0.0) {
    rect.x += rect.width;
    rect.width = -rect.width;
  }
  if (rect.height < 0.0) {
    rect.y += rect.height;
    rect.height = -rect.height;
  }
  return rect;
}

ShaulaRect shaula_rect_clamped(ShaulaRect input, double max_width,
                               double max_height) {
  ShaulaRect rect = shaula_rect_normalized(input);
  double x1 = shaula_clamp_double(rect.x, 0.0, max_width);
  double y1 = shaula_clamp_double(rect.y, 0.0, max_height);
  double x2 = shaula_clamp_double(rect.x + rect.width, 0.0, max_width);
  double y2 = shaula_clamp_double(rect.y + rect.height, 0.0, max_height);
  return shaula_rect_from_points((ShaulaPoint){x1, y1}, (ShaulaPoint){x2, y2});
}

ShaulaRect shaula_rect_expanded(ShaulaRect input, double amount) {
  ShaulaRect rect = shaula_rect_normalized(input);
  return (ShaulaRect){rect.x - amount, rect.y - amount,
                      rect.width + amount * 2.0, rect.height + amount * 2.0};
}

ShaulaRect shaula_rect_union(ShaulaRect a_input, ShaulaRect b_input) {
  ShaulaRect a = shaula_rect_normalized(a_input);
  ShaulaRect b = shaula_rect_normalized(b_input);
  double x1 = MIN(a.x, b.x);
  double y1 = MIN(a.y, b.y);
  double x2 = MAX(a.x + a.width, b.x + b.width);
  double y2 = MAX(a.y + a.height, b.y + b.height);
  return (ShaulaRect){x1, y1, x2 - x1, y2 - y1};
}

gboolean shaula_rect_is_empty(ShaulaRect input) {
  ShaulaRect rect = shaula_rect_normalized(input);
  return rect.width <= 0.5 || rect.height <= 0.5;
}

gboolean shaula_rect_contains_point(ShaulaRect input, ShaulaPoint point) {
  ShaulaRect rect = shaula_rect_normalized(input);
  return point.x >= rect.x && point.y >= rect.y &&
         point.x <= rect.x + rect.width && point.y <= rect.y + rect.height;
}

gboolean shaula_rect_intersects(ShaulaRect a_input, ShaulaRect b_input) {
  ShaulaRect a = shaula_rect_normalized(a_input);
  ShaulaRect b = shaula_rect_normalized(b_input);
  return a.x <= b.x + b.width && a.x + a.width >= b.x &&
         a.y <= b.y + b.height && a.y + a.height >= b.y;
}

double shaula_point_distance(ShaulaPoint a, ShaulaPoint b) {
  double dx = a.x - b.x;
  double dy = a.y - b.y;
  return sqrt(dx * dx + dy * dy);
}

double shaula_point_distance_to_segment(ShaulaPoint point, ShaulaPoint a,
                                        ShaulaPoint b) {
  double dx = b.x - a.x;
  double dy = b.y - a.y;
  double length_squared = dx * dx + dy * dy;
  if (length_squared <= 0.000001)
    return shaula_point_distance(point, a);

  double projection =
      ((point.x - a.x) * dx + (point.y - a.y) * dy) / length_squared;
  double t = shaula_clamp_double(projection, 0.0, 1.0);
  ShaulaPoint projected = {a.x + t * dx, a.y + t * dy};
  return shaula_point_distance(point, projected);
}

ShaulaPoint shaula_point_clamped(ShaulaPoint point, double max_width,
                                 double max_height) {
  return (ShaulaPoint){shaula_clamp_double(point.x, 0.0, max_width),
                       shaula_clamp_double(point.y, 0.0, max_height)};
}

File: src/preview/preview_geometry.zig
const std = @import("std");

const CInt = c_int;
const TRUE: CInt = 1;
const FALSE: CInt = 0;

const ShaulaPoint = extern struct {
    x: f64,
    y: f64,
};

const ShaulaRect = extern struct {
    x: f64,
    y: f64,
    width: f64,
    height: f64,
};

const ShaulaColor = extern struct {
    r: f64,
    g: f64,
    b: f64,
    a: f64,
};

export fn shaula_color_default() ShaulaColor {
    return .{ .r = 0.165, .g = 0.290, .b = 0.400, .a = 1.0 };
}

export fn shaula_color_to_hex(color: ShaulaColor, out: [*]u8) void {
    const r = colorChannel(color.r);
    const g = colorChannel(color.g);
    const b = colorChannel(color.b);
    _ = std.fmt.bufPrint(out[0..8], "#{X:0>2}{X:0>2}{X:0>2}", .{ r, g, b }) catch {};
}

export fn shaula_rect_from_points(a: ShaulaPoint, b: ShaulaPoint) ShaulaRect {
    return shaula_rect_normalized(.{ .x = a.x, .y = a.y, .width = b.x - a.x, .height = b.y - a.y });
}

export fn shaula_rect_normalized(input: ShaulaRect) ShaulaRect {
    var rect = input;
    if (rect.width < 0) {
        rect.x += rect.width;
        rect.width = -rect.width;
    }
    if (rect.height < 0) {
        rect.y += rect.height;
        rect.height = -rect.height;
    }
    return rect;
}

export fn shaula_rect_clamped(input: ShaulaRect, max_width: f64, max_height: f64) ShaulaRect {
    const rect = shaula_rect_normalized(input);
    const x1 = std.math.clamp(rect.x, 0.0, max_width);
    const y1 = std.math.clamp(rect.y, 0.0, max_height);
    const x2 = std.math.clamp(rect.x + rect.width, 0.0, max_width);
    const y2 = std.math.clamp(rect.y + rect.height, 0.0, max_height);
    return shaula_rect_from_points(.{ .x = x1, .y = y1 }, .{ .x = x2, .y = y2 });
}

export fn shaula_rect_expanded(input: ShaulaRect, amount: f64) ShaulaRect {
    const rect = shaula_rect_normalized(input);
    return .{ .x = rect.x - amount, .y = rect.y - amount, .width = rect.width + amount * 2.0, .height = rect.height + amount * 2.0 };
}

export fn shaula_rect_union(a_input: ShaulaRect, b_input: ShaulaRect) ShaulaRect {
    const a = shaula_rect_normalized(a_input);
    const b = shaula_rect_normalized(b_input);
    const x1 = @min(a.x, b.x);
    const y1 = @min(a.y, b.y);
    const x2 = @max(a.x + a.width, b.x + b.width);
    const y2 = @max(a.y + a.height, b.y + b.height);
    return .{ .x = x1, .y = y1, .width = x2 - x1, .height = y2 - y1 };
}

export fn shaula_rect_is_empty(input: ShaulaRect) CInt {
    const rect = shaula_rect_normalized(input);
    return if (rect.width <= 0.5 or rect.height <= 0.5) TRUE else FALSE;
}

export fn shaula_rect_contains_point(input: ShaulaRect, point: ShaulaPoint) CInt {
    const rect = shaula_rect_normalized(input);
    const contains = point.x >= rect.x and point.y >= rect.y and point.x <= rect.x + rect.width and point.y <= rect.y + rect.height;
    return if (contains) TRUE else FALSE;
}

export fn shaula_rect_intersects(a_input: ShaulaRect, b_input: ShaulaRect) CInt {
    const a = shaula_rect_normalized(a_input);
    const b = shaula_rect_normalized(b_input);
    const intersects = a.x <= b.x + b.width and a.x + a.width >= b.x and a.y <= b.y + b.height and a.y + a.height >= b.y;
    return if (intersects) TRUE else FALSE;
}

export fn shaula_point_distance(a: ShaulaPoint, b: ShaulaPoint) f64 {
    const dx = a.x - b.x;
    const dy = a.y - b.y;
    return @sqrt(dx * dx + dy * dy);
}

export fn shaula_point_distance_to_segment(point: ShaulaPoint, a: ShaulaPoint, b: ShaulaPoint) f64 {
    const dx = b.x - a.x;
    const dy = b.y - a.y;
    const len2 = dx * dx + dy * dy;
    if (len2 <= 0.000001) return shaula_point_distance(point, a);
    const t = std.math.clamp(((point.x - a.x) * dx + (point.y - a.y) * dy) / len2, 0.0, 1.0);
    const projection = ShaulaPoint{ .x = a.x + t * dx, .y = a.y + t * dy };
    return shaula_point_distance(point, projection);
}

export fn shaula_point_clamped(point_input: ShaulaPoint, max_width: f64, max_height: f64) ShaulaPoint {
    return .{
        .x = std.math.clamp(point_input.x, 0.0, max_width),
        .y = std.math.clamp(point_input.y, 0.0, max_height),
    };
}

fn colorChannel(value: f64) u8 {
    return @intFromFloat(std.math.clamp(value, 0.0, 1.0) * 255.0 + 0.5);
}
